packages/suite/src/middlewares/wallet/coinjoinMiddleware.ts

```diff
@@ -164,15 +88,77 @@ export const coinjoinMiddleware =
             }
         }
 
+        // Pause coinjoin session when device disconnects.
+        // This is not treated a temporary interruption with automatic restore because the user probably disconnects the device willingly.
+        if (action.type === DEVICE.DISCONNECT && action.payload.id) {
+            api.dispatch(coinjoinAccountActions.pauseCoinjoinSessionByDeviceId(action.payload.id));
+        }
+
+        // Pause/restore coinjoin session when Suite goes offline/online.
+        // This is just UX improvement as the session could not continue offline anyway.
+        if (action.type === SUITE.ONLINE_STATUS) {
+            if (action.payload === false) {
+                api.dispatch(coinjoinAccountActions.pauseInterruptAllCoinjoinSessions());
+            } else if (action.payload === true) {
+                api.dispatch(coinjoinAccountActions.restoreInterruptedCoinjoinSessions());
+            }
+        }
+
+        // Pause/restore coinjoin session based on Tor status.
+        // Continuing coinjoin would be a privacy risk.
         if (action.type === SUITE.TOR_STATUS) {
             if (['Disabling', 'Disabled', 'Error'].includes(action.payload)) {
                 api.dispatch(coinjoinAccountActions.pauseInterruptAllCoinjoinSessions());
+            } else if (action.payload === 'Enabled') {
+                api.dispatch(coinjoinAccountActions.restoreInterruptedCoinjoinSessions());
             }
-            // We restore sessions that were interrupted when successfully Enabled if
-            // there is not any other condition blocking coinjoin resume.
+        }
+
+        // Pause/restore coinjoin session when an account goes out of sync or in sync.
+        // As this is not crucial, it does not pause during the critical phase not to ruin a round.
+        if (accountsActions.endCoinjoinAccountSync.match(action)) {
             const state = api.getState();
-            if (action.payload === 'Enabled' && !isCoinjoinSessionBlockedGlobally(state)) {
-                restoreInterruptedCoinjoinSessions(state);
+            const { accountKey, status } = action.payload;
+            const session = selectCoinjoinAccountByKey(state, accountKey)?.session;
+            if (status === 'out-of-sync' && session && !session?.paused && !session?.starting) {
+                const isAccountInCriticalPhase =
+                    selectIsAccountWithSessionInCriticalPhaseByAccountKey(state, accountKey);
+                if (!isAccountInCriticalPhase) {
+                    api.dispatch(coinjoinAccountActions.pauseCoinjoinSession(accountKey, true));
+                }
+            } else if (status === 'ready' && session?.interrupted) {
+                const account = selectAccountByKey(state, accountKey);
+                if (account) {
+                    const blockers = selectCoinjoinSessionBlockersByAccount(state, account);
+                    if (!Object.values(blockers).some(value => value))
```

Comment:
Check for `isAccountInCriticalPhase` to not fire an unnecessary action?

Reply:
I think the check for `isAccountInCriticalPhase` was originally added by @karliatto to avoid _pausing_ the session when the account is in critical phase because though it is out of sync, it can probably still participate in the round and thus not ruin it.
This logic does not apply to restoring the session and it may even be restored during the critical phase.

Reply:
Yeah, I thought since it cannot be paused because of Tor perhaps it should not be resumed because of Tor :D But it doesn't matter much.